### PR TITLE
reconcile-cluster-role-bindings command

### PR DESCRIPTION
**Pull request by liggitt**
Needed for the updated role binding needed for the existing kubelet client user the master is using

### CODE REVIEW

**Comment by liggitt** [comment]
@deads2k PTAL

**Comment by liggitt** [comment]
@sdodson we'll want to run this command (with `--confirm=true --additive-only=true`) as part of upgrades as well, after reconciling roles

File: pkg/cmd/admin/policy/reconcile_clusterrolebindings.go [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+	// extra subjects only require a change if we're not unioning
+	if len(extraActualSubjects) > 0 && !union {
+		needsUpdating = true
+	}
+
+	if !needsUpdating {
+		return nil, false
+	}
+
+	var updated authorizationapi.ClusterRoleBinding = *expected
```
Comment (deads2k) [inline pkg/cmd/admin/policy/reconcile_clusterrolebindings.go]: labels and annotations from the actual.

Comment (deads2k) [inline pkg/cmd/admin/policy/reconcile_clusterrolebindings.go]: > labels and annotations from the actual.

I see it.  Add godoc indicating that not all fields in this new returned `ClusterRoleBinding` match the actual.  Only respect certain fields.

Comment (deads2k) [inline pkg/cmd/admin/policy/reconcile_clusterrolebindings.go]: > I see it. Add godoc indicating that not all fields in this new returned ClusterRoleBinding match the actual. Only respect certain fields.

No, you use this for display.  Set it properly.

Comment (liggitt) [inline pkg/cmd/admin/policy/reconcile_clusterrolebindings.go]: same issue in reconcile-cluster-roles... should we stomp for now assuming the bootstrap ones don't set labels/annotations?

File: pkg/cmd/admin/policy/reconcile_clusterrolebindings.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+			continue
+		}
+
+		// RoleRef is immutable, to reset this, we have to delete/recreate
+		if !kapi.Semantic.DeepEqual(roleBinding.RoleRef, changedRoleBindings[i].RoleRef) {
+			roleBinding.RoleRef = changedRoleBindings[i].RoleRef
+			roleBinding.Subjects = changedRoleBindings[i].Subjects
+
+			// TODO: for extra credit, determine whether the right to delete/create this rolebinding for the current user came from this rolebinding before deleting it
+			err := o.RoleBindingClient.Delete(roleBinding.Name)
```
Comment (deads2k) [inline pkg/cmd/admin/policy/reconcile_clusterrolebindings.go]: You should update when possible.  You only need to delete/create when you're changing a roleRef, right?

File: pkg/cmd/admin/policy/reconcile_clusterrolebindings.go (hunk 1 above)
Comment (deads2k) [inline pkg/cmd/admin/policy/reconcile_clusterrolebindings.go]: > same issue in reconcile-cluster-roles... should we stomp for now assuming the bootstrap ones don't set labels/annotations?

Fix this one, make @stevekuznetsov an issue to fix the other one.

File: pkg/cmd/admin/policy/reconcile_clusterrolebindings.go (hunk 2 above)
Comment (liggitt) [inline pkg/cmd/admin/policy/reconcile_clusterrolebindings.go]: yes... which is why this is in a block that checks roleRef equality...

Comment (deads2k) [inline pkg/cmd/admin/policy/reconcile_clusterrolebindings.go]: > You should update when possible. You only need to delete/create when you're changing a roleRef, right?

It's clearly too late to be doing this review.  Retracted.

**Comment by deads2k** [comment]
lgtm

File: pkg/cmd/admin/policy/reconcile_clusterrolebindings.go (hunk 1 above)
Comment (liggitt) [inline pkg/cmd/admin/policy/reconcile_clusterrolebindings.go]: updated in the calling func, with copious comments. also fixed reconcile-roles and added tests

**Comment by liggitt** [comment]
[test]

**Comment by liggitt** [comment]
hmm.. need to split out the default bindings for "all authenticated"... probably don't want to re-add those

**Comment by deads2k** [comment]
> hmm.. need to split out the default bindings for "all authenticated"... probably don't want to re-add some of those automatically (specifically the "self-provisioner" role binding)

I'm inclined to keep the behavior simple.  Replace all.  I could see adding a `[]string` flag to blacklist certain bootstrap rolebindings and roles.  I'd prefer that as a followup that updates both.

**Comment by deads2k** [comment]
lgtm. merge when you feel happy about it.

**Comment by liggitt** [comment]
> I'm inclined to keep the behavior simple. Replace all. I could see adding a []string flag to blacklist certain bootstrap rolebindings and roles. I'd prefer that as a followup that updates both

Not sure what I'd set that to for the case where the upgrade playbook automatically runs these commands

**Comment by deads2k** [comment]
> Not sure what I'd set that to for the case where the upgrade playbook automatically runs these commands

Changing permissions is always fraught.  I suppose I could also see a flag that prevents certain subjects from getting re-bound to any role, but I don't want something in this command logic to make that impossible to do.

**Comment by liggitt** [comment]
Added `--do-not-add-groups` and `--do-not-add-users` (name suggestions welcome), which prevents adding new role-bindings but does not remove existing ones.

My recommended automatic invocation would probably be something like this:
```
oadm policy reconcile-cluster-role-bindings \
  --do-not-add-groups=system:authenticated \
  --do-not-add-users=system:anonymous \
  --additive-only=true --confirm
```

That would let us avoid regranting permissions like request-project if the admin had specifically removed them.

**Comment by deads2k** [comment]
> Added --do-not-add-groups and --do-not-add-users (name suggestions welcome), which prevents adding new role-bindings but does not remove existing ones.

`--excluded-users`?

**Comment by liggitt** [comment]
I started with `--exclude-users` and `--exclude-groups`, but wondered if that was clear that means those are never added, but also not removed if they exist both in the expected binding and actual binding. Shorten the name and doc it?

**Comment by deads2k** [comment]
> I started with --exclude-users and --exclude-groups, but wondered if that was clear that means those are never added, but also not removed if they exist both in the expected binding and actual binding. Shorten the name and doc it?

Yeah.  The option is going to be difficult to explain no matter what its named.

**Comment by deads2k** [comment]
Ran afoul of one pet peeve: reassigning parameter reference.  Its particularly insidious with slices, since slices are passed by reference, so you can mutate the content, but reassigning that reference and then mutating is a safe operation.  That makes it safe, but I don't like having to think about it.

Otherwise, lgtm.

**Comment by liggitt** [comment]
[merge]

**Pull request merged**